### PR TITLE
gh-84852: Add MIME types for .eot, ,otf, .ttf, .woff and .woff2 fonts

### LINKED ISSUE #84852: Add missing mimetypes for fonts
BPO | [40675](https://bugs.python.org/issue40675)
--- | :---
Nosy | @sahil290791
PRs | <li>python/cpython#20199</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2020-05-19.01:08:13.333>
labels = ['library']
title = 'Add missing mimetypes for fonts'
updated_at = <Date 2020-05-19.01:08:13.333>
user = 'https://github.com/sahil290791'
```

bugs.python.org fields:
```python
activity = <Date 2020-05-19.01:08:13.333>
actor = 'sahil.prajapati'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2020-05-19.01:08:13.333>
creator = 'sahil.prajapati'
dependencies = []
files = []
hgrepos = []
issue_num = 40675
keywords = []
message_count = 1.0
messages = ['369312']
nosy_count = 1.0
nosy_names = ['sahil.prajapati']
pr_nums = ['20199']
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = None
url = 'https://bugs.python.org/issue40675'
versions = []
```

</p></details>

### Linked PRs
* gh-113592

### PR DESCRIPTION
**Pull request by sahil290791**
# Add missing font mime types

Fixes https://github.com/python/cpython/issues/84852.

Currently the guess_type method doesnt work for fonts of type: .eot, .woff, .woff2, .otf and .ttf
I have added the neccessary mimetypes for these fonts

Reference: https://codingstill.com/2013/01/set-mime-types-for-web-fonts-in-iis/

```
[bpo-40675](https://bugs.python.org/issue40675): Summary of the changes made
```

* Issue: gh-84852

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

## Recognized GitHub username

We couldn't find a [bugs.python.org](https://bugs.python.org/) (b.p.o) account corresponding to the following GitHub usernames:

@sahil290791

This might be simply due to a missing "GitHub Name" entry in one's b.p.o account settings. This is necessary for legal reasons before we can look at this contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.




You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

File: Lib/mimetypes.py [hunk 1]
```diff
@@ -543,6 +543,10 @@ def _default_mime_types():
         '.webm'   : 'video/webm',
         '.avi'    : 'video/x-msvideo',
         '.movie'  : 'video/x-sgi-movie',
+        '.woff'   : 'application/font-woff',
```
Comment (tiran) [inline Lib/mimetypes.py]: The dict is more or less sorted by value. Please move the new 'application' mimetypes up.

**Review by tiran — DISMISSED** [review]

File: Lib/mimetypes.py (hunk 1 above)
Comment (sahil290791) [reply Lib/mimetypes.py]: Hi @tiran, I have sorted it based on values.

**Comment by csabella** [comment]
@sahil290791, please sign the CLA as per the comment from @the-knights-who-say-ni bot.  Thanks!

**Comment by sahil290791** [comment]
@csabella I have signed the CLA now, but I guess it is going to take 2-3 days before it starts reflecting. I will update on this PR once it starts reflecting. Thanks 👍

**Comment by sahil290791** [comment]
@csabella the CLA has been signed and is reflecting in the PR.

**Comment by sahil290791** [comment]
I have made the requested changes; please review again.

**Comment by sahil290791** [comment]
@tiran @csabella  Please let me know if anything else is pending from my side?

It would be great if this can be merged as currently I am working by adding certain if conditions in my current project.

**Comment by azimovMichael** [comment]
Hi, @tiran @csabella any estimation on if and when this will be merged?
Also, @sahil290791, you don't need if conditions, you can add these types dynamically with:
`mimetypes.add_type('application/font-woff2', '.woff2')`

**Comment by sahil290791** [comment]
Thanks @azimovMichael for the suggestion, will use dynamic approach currently.

**Comment by MinchinWeb** [comment]
In 2017, the IANA approved a "fonts" top level media type, and so the official MIME types are now:

Name | Template | Reference
-- | -- | --
collection | font/collection | [RFC8081](https://www.rfc-editor.org/rfc/rfc8081.html)
otf | font/otf | [RFC8081]
sfnt | font/sfnt | [RFC8081]
ttf | font/ttf | [RFC8081]
woff | font/woff | [RFC8081]
woff2 | font/woff2 | [RFC8081]

[source](https://www.iana.org/assignments/media-types/media-types.xhtml#font)

**Comment by hugovk** [comment]
I've updated this to fix the merge conflict, use the RFC 8081 MIME types, corrected the Embedded OpenType (EOT) MIME type (append `object`: `application/vnd.ms-fontobject`), and added a What's New entry.

File: Lib/test/test_mimetypes.py [hunk 2]
```diff
@@ -227,11 +227,16 @@ def check_extensions():
             self.assertEqual(mimetypes.guess_extension('application/postscript'), '.ps')
             self.assertEqual(mimetypes.guess_extension('application/vnd.apple.mpegurl'), '.m3u')
             self.assertEqual(mimetypes.guess_extension('application/vnd.ms-excel'), '.xls')
+            self.assertEqual(mimetypes.guess_extension('application/vnd.ms-fontobject'), '.eot')
```
Comment (colesbury) [inline Lib/test/test_mimetypes.py]: It looks like this is going to conflict with the refactoring in https://github.com/python/cpython/pull/126412.

**Review by colesbury — APPROVED** [review]

File: Lib/mimetypes.py [hunk 3]
```diff
@@ -542,6 +543,10 @@ def _default_mime_types():
         '.aiff'   : 'audio/x-aiff',
         '.ra'     : 'audio/x-pn-realaudio',
         '.wav'    : 'audio/x-wav',
+        '.otf'    : 'font/otf',
+        '.ttf'    : 'font/ttf',
+        '.woff'   : 'font/woff',
+        '.woff2'  : 'font/woff2',
```
Comment (terryjreedy) [inline Lib/mimetypes.py:550]: I verified from the RFC that these 4 lines and hence the corresponding test lines are correct.  There is no reference I see for the vnd-ms addition so cannot verify.  The merge conflict will be in the tests due to the switch to using subtests, which looks good to me.  So merge that first and fix this.

Comment (hugovk) [reply Lib/mimetypes.py:550]: Thanks for verifying.

`application/vnd.ms-fontobject` is vendor specific and can be seen for example at:

* https://www.iana.org/assignments/media-types/application/vnd.ms-fontobject
* https://developer.mozilla.org/en-US/docs/Web/HTTP/MIME_types/Common_types

File: Doc/whatsnew/3.14.rst [hunk 4]
```diff
@@ -395,7 +405,6 @@ multiprocessing
 
   (Contributed by Roy Hyunjin Han for :gh:`103134`)
 
-
 operator
```
Comment (hugovk) [inline Doc/whatsnew/3.14.rst]: Unintentional line delete from merge, adding it back:
```suggestion

operator
```

**Comment by hugovk** [comment]
Thank you @sahil290791 for the initial PR and everyone else for the reviews!

**Comment by sahil290791** [comment]
Thanks @hugovk and the reviewers for getting this reviewed and merged. 😃

**Pull request merged**